### ISSUE TITLE
lnwallet doesn't see channels as closed; attempts to close them are rejected as double-spends

### ISSUE THREAD

**Issue by aakselrod** [lightningnetwork/lnd#142]
`mably` and `juscamarena` on Slack are reporting issues closing channels. The closing transaction is seen (correctly) as a double-spend. Examples below:

```
-22: TX rejected: transaction b68ad8dc0c7e0bf5b59699b1c117bf4687f2220e787ca15f396da0ba07fb1a09:0 tried to double spend output e987873ac689c4035c54ba1f7ffb78f062bf8ba26ff7e0b9839ac04a561959b3:0
```

```
-22: TX rejected: transaction d7b14223bed608d9a46444ecf87562d741f08f0f327da8f63bd78becf9d5b114:0 tried to double spend output a5a5cc92bb8bb98c27b59dd35de2c705db78a59cc8a4b85c28ae2926a1320c7b:0
```

This is correct and the channels were closed earlier, but the wallet didn't see it.

I believe @Roasbeef is already working on the fix with a state machine update; either way, this issue is to track that work and/or a fix.

Another means of mitigating issues like this might be to create a `rescan` RPC/CLI command.

**Comment by mably** [comment]
I'll just add that these ghost channels seems to stall any sendpayment command you make even if you have opened another similar channel.

**Comment by aakselrod** [comment]
I'm still working on reproducing this, but have an additional example with a bit more detail from `moli` on Slack:

```
C:\Users\moli>lncli closechannel --funding_txid=de664127bf62bdd833c42260707390d130d73067a48e34ac78008381af9185ab --output_index=0 --force rpc error: code = 2 desc = -22: TX rejected: transaction e12e9aabaac4fb1c9949486de5a649fd908da53a140876e83ac830f2e20acb3d:0 tried to double spend output de664127bf62bdd833c42260707390d130d73067a48e34ac78008381af9185ab:0
```

and

```
C:\Users\moli>lncli listchannels

{

    "channels": [

        {

            "remote_pubkey": "0280c83b3eded413dcec12f7952410e2738f079bd9cbc9a7c462e32ed4d74bd5b7",

            "channel_point": "de664127bf62bdd833c42260707390d130d73067a48e34ac78008381af9185ab:0",

            "chan_id": "0",

            "capacity": "205000",

            "local_balance": "12380",

            "remote_balance": "187614",

            "unsettled_balance": "0",

            "total_satoshis_sent": "0",

            "total_satoshis_received": "2399",

            "num_updates": "2619",

            "pending_htlcs": [

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                },

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                },

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                },

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                },

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                },

                {

                    "incoming": true,

                    "amount": "1",

                    "hash_lock": "cs1uhCLEB/ttCYaQ8RMLfe1+wvf14dML2dUh8BU2N5M=",

                    "expiration_height": 0,

                    "revocation_delay": 0

                }

            ]

        },
...
```

**Comment by Roasbeef** [comment]
The issue is is mishandling of the `max_in_flight_htlc` channel policy paramter by the current `htlcManager` goroutine. Due to jittery latency between hops, when performing a large number of consistent streaming payments, it's possible that the number of uncommitted changes `state.pendingBatch` exceeds the `max_in_flight_htlc` policy paramter for that particular channel. When this event happens the `AddHTLC` method on the `lnwallet.LightningChannel` object returns an error as the the commitment sanity has been violated by the local caller by attempting to either add a new HTLC to the update log (which causes an overflow), or initiate a new state transition with a commitment transaction that exceeds the maximum number of HTLC's. 

In this case, (with the current logic) the `htlcManager` will attempt to abandon the connection as a policy has been violated. This abandonment leads to desynchronization in the commitment states as message retransmission is currently unimplemented. Upon disconnect, due to the fact that we attempt to maintain a persistent connection to all peers we have a channel open with, a new connection will be reestablished (ignoring the possibly inflight updates) and the whole process starts again with the commitment states getting further and further out of sync. 

Once the channel is in this state, neither side is able to complete a state transition as they have different views of  both the local and remote commitment chain. As shown from the `ListChannels` result above, both commitment states end up having many unsettled HTLC's which are unable to be properly settled. In the case that either side attempts to force close the channel, the other side won't detect that force close as it's doesn't realize the states are out of sync. 

The fix for the series of issues described above is the following: 
 * Modify the `closeObserver` goroutine that's launched by a `LightingChannel` once initialized to instead track if it's a remote or local broadcast once again by the commitment transaction itself. This was the vase before, but was changed when it was realize that there existed a race condition between the `breachAbriter` and a new peer coming online and initiates a state transition. This was fixed by commit 2c0d5e0f0d253073016a1693c7a06d21e6e0f06f. 
    * The force close is being undetected as it's currently hitting the last case in the switch statement which is considered "impossible" under a sane assumption that the states never desynchronize. 
 * Modify the logic surrounding the HTLC batch and trickle logic within the `htlcManager` to *ensure* that the size of `state.pendingBatch` never exceeds the current `max_in_flight_htlc` policy for the channel that it's managing. 
     * This can be implemented by adding a distinct overflow bin that's populated once `len(state.pendingBatch) >= channel.MaxInFlightHTLCs`. Upon the next state transition, this bin should be emptied immddatly before attempting to obtain any new HTLC's from the downstream channel connected to the `htlcSwitch`. 
       * In other words, the following prioritization needs to be in place: 
            1. overflow HTLC's
            2. HTLC's in the pending batch
            3. new HTLC's from the downstream switch channel

**Comment by Roasbeef** [comment]
The overflow bug was fixed by: 505421db2c78500802194546076588eb17cb297b. 

Closing
